### PR TITLE
Enhancement for categories management and products grids

### PR DESCRIPTION
**Pull request by luigifab**
### Description

It allow to keep current selected tab on store switch, and to keep current category and current store on refresh in categories management. It prevents to add a new category without a parent category.

It add columns to products grid in categories (status, visibility, action).

image

It open product edit in a popin (without header and footer) with the current selected store in categories and in associated products of configurable products. You can also open the edit page in a new tab or in a new window.

image

It allow to check or uncheck products only when clicking on the checkbox or in the td that contain the checkbox for all grids (products in categories ; related, upsell, crosssell, associated (for grouped and configurable products)).

OpenMage 20.0.13 / PHP 7.4.11 - 8.0.11

### Manual testing scenarios

**TEST 1**
Go to Catalog / Categories.

- Try to create a new subcategory without selecting a parent. This is not possible.
- You can check or uncheck products only when clicking on td input.
- New columns "Status", "Visibility" and "Action".
- The edit link opens product edit page in popup (with the current store).
- You can open the [Edit] link in a new browser tab (middle click or right click "Open in a new tab/window").
- Use Name, Price, Status and Visibility values from selected store.

**TEST 2**
Go to Catalog > Categories the select any category.
Go to Category Products tab.

- Select another store with store switcher.  The current tab is not lost.
- Press F5 or refresh the page. The current category and the current store are not lost.
- Uncheck all products then reset filter or filter or sort or page change. After AJAX all products are unchecked.

**TEST 3**
Go to Catalog > Products and choose any configurable product. For Associated Products grid.

- You can check or uncheck products only when clicking on td input.
- The [Edit] link opens product edit page in popup (with the current store)
- You can open the [Edit] link on a new tab (middle click or right click open in a new tab/window)

**TEST 4**
Go to Catalog > Products and choose any product.
For Related, Up-sells and Cross-sells, Associated (for grouped products) grids.

- Uncheck all products then reset filter or filter or sort or page change. After AJAX all products are unchecked.
- Check some products then reset filter or filter or sort or page change. After AJAX selected products are checked.
- You can check check or uncheck products only when clicking on td input.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)
 - [x] Add yourself to contributors list

### CODE REVIEW

**Review by colinmollenhour — DISMISSED** [review]

**Comment by ADDISON74** [comment]
Could you please provide some insides why you wanted to add those columns? For example Status column - it could be OK. In this way you know what products are Enable/Disable.

My suggestion is for a link at the end of the row to edit the product. This could be a much better option. Once I see a product Disable I can edit it with one click.

**Comment by luigifab** [comment]
Yes, this is only to easily see which products can be displayed in the category.
We can also add a link to the product page, why not.

**Comment by Flyingmana** [comment]
Could someone check the sql queries which change here?
Having a `Explain` of the query would be even better.

**Comment by luigifab** [comment]
Request before:

```sql
SELECT `e`.*, `at_position`.`position` FROM `catalog_product_entity` AS `e`
 LEFT JOIN `catalog_category_product` AS `at_position` ON (at_position.`product_id`=e.entity_id) AND (category_id=2)
```
image

Request after;
```sql
SELECT `e`.*, `at_position`.`position`, `at_name`.`value` AS `name`, `at_price`.`value` AS `price`, `at_status`.`value` AS `status`, `at_visibility`.`value` AS `visibility` FROM `catalog_product_entity` AS `e`
 LEFT JOIN `catalog_category_product` AS `at_position` ON (at_position.`product_id`=e.entity_id) AND (category_id=2)
 LEFT JOIN `catalog_product_entity_varchar` AS `at_name` ON (`at_name`.`entity_id` = `e`.`entity_id`) AND (`at_name`.`attribute_id` = '73') AND (`at_name`.`store_id` = 0)
 LEFT JOIN `catalog_product_entity_decimal` AS `at_price` ON (`at_price`.`entity_id` = `e`.`entity_id`) AND (`at_price`.`attribute_id` = '77') AND (`at_price`.`store_id` = 0)
 LEFT JOIN `catalog_product_entity_int` AS `at_status` ON (`at_status`.`entity_id` = `e`.`entity_id`) AND (`at_status`.`attribute_id` = '98') AND (`at_status`.`store_id` = 0)
 LEFT JOIN `catalog_product_entity_int` AS `at_visibility` ON (`at_visibility`.`entity_id` = `e`.`entity_id`) AND (`at_visibility`.`attribute_id` = '104') AND (`at_visibility`.`store_id` = 0)
```
image

**Comment by luigifab** [comment]
I added a link to the product page. Click on it with the mouse wheel.

**Comment by ADDISON74** [comment]
@luigifab - A link to edit a product from Categories page is an nice feature. Please take in consideration these two issues:

- **Visibility** of the feature. It shouldn't be a mouse wheel click because if you use a laptop there are situations when you couldn't have a mouse device available and it will become a problem. My suggestion is to add a new column at the right end. Inside every cell insert an Edit link. It is more intuitive and reveals the scope of the action.

- **Target** for the link. Opening the product edit page in the same window or in a new window? I would choose opening in a new window,  the same behavior like when editing Associated Products in configurable products. It is a new small window and after pressing the [Save] button appears a new content with the [Close] button at the bottom. If you don't press the close button the window will close automatically. In this way you stay in Categories page without coming back after editing the product. Why editing a product? From my experience you could have a product Out of Stock or Disable and even it is listed in the grid it won't appear in Frontend.

**Comment by luigifab** [comment]
You are right, I doing that.

**Comment by luigifab** [comment]
I added the magic link. Not really tested for now.

**Comment by ADDISON74** [comment]
Nice addition. I will test this PR. 

Recently in this page dedicated to categories I was bothered to deselect the products when clicking on a row. I accidentally deselected several products just because I clicked next to the position edit input box. The select/unselect actions must be done only on clicking the checkbox at the beginning of the row, not on the whole row.

**Comment by luigifab** [comment]
I replaced the popup window by a popup iframe.
I can easily use this popup for associated products edit link in configurable products.

The select/unselect actions is now done only on clicking the checkbox or in the cell of the checkbox.

File: app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php
Comment (luigifab) [inline app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php:283]: I don't understand PHPStan here...

**Review by fballiano — APPROVED** [review]

File: app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php
Comment (sreichel) [reply app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php:289]: ```$this->getLayout()->getBlock('root');``` returns Mage_Core_Block_Abstract, but ```addBodyClass()``` belongs to a child class ... this should work:

```
            /** @var Mage_Adminhtml_Block_Page $root */
            $root = $this->getLayout()->getBlock('root');
            if ($root) {
```

**Comment by colinmollenhour** [comment]
I skimmed the diff and it looks good but have not done any testing. I think someone besides me who uses these features more often than I should take a look such as @ADDISON74 .

**Comment by ADDISON74** [comment]
@colinmollenhour - I will provide my feedback after testing this PR.

Please take a look to this comment to merge that PR https://github.com/OpenMage/magento-lts/pull/2349#issuecomment-1201785729

**Comment by ADDISON74** [comment]
I made a change to the initial post to better see the tests that need to be done.

**FOR TEST 1**
1. A subcategory cannot be created if the parent category is not selected first. I would change the message displayed in the popup window because it is not well formulated ( "Please select a parent category before add new one.")

2. Grid from the "Category Products" tab.
- I appreciate that selecting/deselecting a row is no longer done by an accidental click on the row area, but only in the box on the first column.

- The "Action" column is welcome. Product editing can be done either in the popup window or in another window, which is a plus. The popup window resembles the one when we edit a simple product from a configurable product. I have not analyzed if they are the same but I would look at the buttons that appear on the page.

- The "Status" column is again a bonus. Many times products were selected in a category and they were not displayed in the Frontend because they were not Enabled. Now it is possible to immediately change the status of the product via [Edit] action. It would have been interesting to apply a Mass Action for selected products.

- "Visibility" column is useless. We find it in Catalog > Products. Its presence in this grid does not give me any advantage, At least at first glance.

This test was not an analysis of the source code, but from the perspective of a user who wants to do everything possible in a page opened in the browser.

**Comment by ADDISON74** [comment]
**TEST 1**
Regarding the grid in the "Category Products" tab. When it is desired to add products to a category, it would be recommended to list the product type on a column, because as it is now we cannot differentiate between a simple product that is not visible in the Frontend and a configurable one. I suggest that the "Visibility" column be replaced with "Type", which is far more useful.

**Comment by ADDISON74** [comment]
I tested this PR and there are issues. unfortunately I will not approve it in the form it is now. It will have to be divided into several PRs so that each one in turn modifies 2 x Category, 1 x Product Configurable Grid and 3 x Product Edit Grid (Related, Up-Sell and Cross-Sells). Each part must be analyzed separately and approved.

This is the result after basic testing and analysis. I used the command ```git fetch origin pull/1379/head:pr1379``` and I got latest OM 1.9.4.x from today August 12th with this PR:

**TEST 1**
Go to Catalog / Categories.

- Try to create a new subcategory without selecting a parent. This is not possible. => **NOT WORKING**
- You can check or uncheck products only when clicking on td input. => **TO BE DISCUSSED**
- New columns "Status", "Visibility" and "Action". => **NEEDS TYPE COLUMN. WIDTH FOR COLUMNS TO FIX ACTION MAINLY**
- The edit link opens product edit page in popup (with the current store). => **NOT WORKING**
- You can open the [Edit] link in a new browser tab (middle click or right click "Open in a new tab/window"). => **OK**
- Use Name, Price, Status and Visibility values from selected store.  => **TO BE DISCUSSED. It seems OM has an issue with this feature  (I was not able to set different prices for English, French, German store views)**

**TEST 2**
Go to Catalog > Categories the select any category.
Go to Category Products tab.

- Select another store with store switcher. The current tab is not lost. => **TO BE DISCUSSED**
- Press F5 or refresh the page. The current category and the current store are not lost. **TO BE DISCUSSED**
- Uncheck all products then reset filter or filter or sort or page change. After AJAX all products are unchecked. **TO BE DISCUSSED**

**TEST 3**
Go to Catalog > Products and choose any configurable product. For Associated Products grid.

- You can check or uncheck products only when clicking on td input. => **NOT WORKING**
- The [Edit] link opens product edit page in popup (with the current store) => **OK**
- You can open the [Edit] link on a new tab (middle click or right click open in a new tab/window) => **OK**

**TEST 4**
Go to Catalog > Products and choose any product.
For Related, Up-sells and Cross-sells, Associated (for grouped products) grids.

- Uncheck all products then reset filter or filter or sort or page change. After AJAX all products are unchecked. => **TO BE DISCUSSED**
- Check some products then reset filter or filter or sort or page change. After AJAX selected products are checked. => **TO BE DISCUSSED**
- You can check check or uncheck products only when clicking on td input. => **NOT WORKING**

**Comment by ADDISON74** [comment]
This is a second test to confirm **NOT WORKING** status from my previous post.

**TEST 1**
Go to **Catalog > Manage Categories**

- Try to create a new subcategory without selecting a parent. This is not possible. => **NOT WORKING**
- The edit link opens product edit page in popup (with the current store). => **NOT WORKING**

**TEST 3**
Go to **Catalog > Manage Products** and edit a configurable product. In **Associated Products** tab grid.

- You can check or uncheck products only when clicking on td input. => **NOT WORKING**

**TEST 4**
Go to **Catalog > Manage Products** and edit a product.
For **Related**, **Up-sells**, **Cross-sells** and Associated Products (for grouped products) tab grids.

- You can check check or uncheck products only when clicking on td input. => **NOT WORKING**

**Comment by ADDISON74** [comment]
I would like to thank @luigifab for this nice work. I tested this PR and there are some issues that need a fix. 

First of all, it has to be divided into several PRs so that each change/feature should be analyzed separately and approved. For example, a PR should be about the "Category Products" grid where 4 columns are entered, the product is edited and the row is selected. Another PR should be linked to the error message in the situation when you want to create a sub-category and you have not selected the parent one. Another PR should be related to the situation switching between the Stores. Here I have practically finished the **Test 1** section.

This is the result after testing and analysis changes and features. I used the command ```git fetch origin pull/1379/head:pr1379``` and I got latest OM 1.9.4.x from today August 13th with this PR:

**TEST 1**
Go to **Catalog > Manage Categories**.

- Try to create a new subcategory without selecting a parent. This is not possible. => **WORKS ONLY AFTER THE PAGE LOADS. If you select a category then keep [Ctrl] key pressed and click on to unselect it then you can add a subcategory, the warning message is not appearing in this case**
- You can check or uncheck products only when clicking on td input. => **TO BE DISCUSSED. We can use click to Edit that product in a popup window, pay attention to mouse pointer which is hand**
- New columns "Status", "Visibility" and "Action". => **NEEDS "TYPE" COLUMN TOO AND SET WIDTHS FOR MOST OF THE COLUMNS TO FIX THE ACTION COLUMN WIDTH**
- The edit link opens product edit page in popup (with the current store). => **OK**
- You can open the [Edit] link in a new browser tab (middle click or right click "Open in a new tab/window"). => **OK**
- Use Name, Price, Status and Visibility values from selected store.  => **TO BE DISCUSSED. This needs more testing, it seems OM has an issue with this feature  (I was not able to set different prices for English, French, German store views)**

**TEST 2**
Go to **Catalog > Manage Categories** then select any category.
Go to **Category Products** tab.

- Select another store with store switcher. The current tab is not lost. => **TO BE DISCUSSED. This needs more testing**
- Press F5 or refresh the page. The current category and the current store are not lost. **TO BE DISCUSSED. This needs more testing**
- Uncheck all products then reset filter or filter or sort or page change. After AJAX all products are unchecked. **TO BE DISCUSSED. We can use click to Edit that product in a popup window, pay attention to mouse pointer which is hand**

**TEST 3**
Go to **Catalog > Manage Products** and choose any configurable product. For Associated Products grid.
- You can check or uncheck products only when clicking on td input. => **TO BE DISCUSSED. We can use click to Edit that product in a popup window, pay attention to mouse pointer which is hand**
- The [Edit] link opens product edit page in popup (with the current store) => **OK**
- You can open the [Edit] link on a new tab (middle click or right click open in a new tab/window) => **OK**

**TEST 4**
Go to **Catalog > Manage Products** and choose any product.
For **Related**, **Up-sells**, **Cross-sells**, **Associated Products** (for Grouped Product type) grids.

- Uncheck all products then reset filter or filter or sort or page change. After AJAX all products are unchecked. => **TO BE DISCUSSED**
- Check some products then reset filter or filter or sort or page change. After AJAX selected products are checked. => **TO BE DISCUSSED**
- You can check/uncheck products only when clicking on td input. => **TO BE DISCUSSED. We can use click to Edit that product in a popup window, pay attention to mouse pointer which is hand.**

**Comment by luigifab** [comment]
I will again check and fix conflicts.

**Review by matteotestoni — DISMISSED** [review]

**Comment by ADDISON74** [comment]
My opinion related to this PR is that it contained too many changes. Ideally it would be for each modification to have its own PR so that the revision is as easy and fast as possible. It can be seen that I tested this PR and identified a lot of problems that need to be solved and discussed. In addition, we still have some conflicts to resolve.

@matteotestoni - out of curiosity, did you test this PR before approving it? With the new RFC such approvals can lead to the integration of a PR that could create issue. Everyone, please test a PR before approving it. We should not invoke that "haste makes waste". Thank you for your understanding.

**Comment by luigifab** [comment]
Yes I started to split in multiple PR.
But in same time I'm using this PR as is.
I rebasing it, and I need to check ADDISON74 comment's.

**Comment by luigifab** [comment]
Closed because this PR must be split in multiple PR... and I will be out there until 2026.
Anyway, I continue use it as is and the version I'm using is [here](https://www.luigifab.fr/diff/3-pull-1379.diff) [1379.diff.txt](https://github.com/OpenMage/magento-lts/files/11610417/3-pull-1379.diff.txt).

**Pull request closed without merging**
